File: RiskAndroid/src/main/java/cc/game/android/risk/DiceDialog.java
package cc.game.android.risk;

import android.app.Dialog;
import android.content.DialogInterface;
import android.view.View;
import android.widget.TextView;

import cc.lib.game.Utils;
import cc.lib.risk.Army;
import cc.lib.utils.Lock;

/**
 * Created by Chris Caron on 9/15/21.
 */
class DiceDialog implements Runnable, DialogInterface.OnDismissListener, DialogInterface.OnClickListener {

    private final RiskActivity context;
    private final Army attacker, defender;
    private final int [] attackingDice, defendingDice;
    private final boolean [] result;
    private final Lock lock;
    private Dialog dialog;
    private TextView [] text;

    public DiceDialog(RiskActivity context, Army attacker, Army defender, int[] attackingDice, int[] defendingDice, boolean[] result) {
        this.lock = new Lock();
        this.context = context;
        this.attacker = attacker;
        this.defender = defender;
        this.attackingDice = attackingDice;
        this.defendingDice = defendingDice;
        this.result = result;
        context.runOnUiThread(this);
        Utils.waitNoThrow(this, 2000);
        if (!dialog.isShowing())
            return;
        lock.block();
        if (!dialog.isShowing())
            return;
        context.runOnUiThread(() -> {
            showResult();
        });
        if (dialog.isShowing()) {
            dialog.setCanceledOnTouchOutside(true);
            Utils.waitNoThrow(this, 6000);
        }
        context.runOnUiThread(() -> dialog.dismiss());
    }

    public void run() {
        View view = View.inflate(context, R.layout.dice_dialog, null);

        TextView tv_attacker = view.findViewById(R.id.tv_attacker);
        TextView tv_defender = view.findViewById(R.id.tv_defender);

        tv_attacker.setText(Utils.toPrettyString(attacker));
        tv_defender.setText(Utils.toPrettyString(defender));

        DiceView [] red = {
                view.findViewById(R.id.red1),
                view.findViewById(R.id.red2),
                view.findViewById(R.id.red3)
        };

        DiceView [] white = {
                view.findViewById(R.id.white1),
                view.findViewById(R.id.white2)
        };

        for (int i=0; i<red.length; i++) {
            if (i >= attackingDice.length) {
                red[i].setVisibility(View.INVISIBLE);
            } else {
                red[i].rollDice(attackingDice[i], lock);
            }
        }

        for (int i=0; i<white.length; i++) {
            if (i >= defendingDice.length)
                white[i].setVisibility(View.INVISIBLE);
            else
                white[i].rollDice(defendingDice[i], lock);
        }

        text = new TextView[] {
                view.findViewById(R.id.text1),
                view.findViewById(R.id.text2),
        };

        dialog = context.newDialogBuilder().setView(view)
                .setNegativeButton("Pause", this).show();
        dialog.setOnDismissListener(this);
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        lock.releaseAll();
        synchronized (this) {
            notify();
        }
    }

    @Override
    public void onClick(DialogInterface dialog, int which) {
        context.stopGameThread();
    }

    void dismiss() {
        dialog.dismiss();
    }

    void showResult() {
        for (int i=0; i<result.length; i++) {
            text[i].setText(result[i] ? R.string.arrow_left : R.string.arrow_right);
        }
    }
}

File: RiskAndroid/src/main/java/cc/game/android/risk/DiceDialog.kt
package cc.game.android.risk

import android.app.Dialog
import android.content.DialogInterface
import android.content.DialogInterface.OnCancelListener
import android.content.DialogInterface.OnDismissListener
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import cc.game.android.risk.databinding.DiceDialogBinding
import cc.lib.risk.Army
import cc.lib.utils.Lock
import cc.lib.utils.prettify

/**
 * Created by Chris Caron on 9/15/21.
 */
internal class DiceDialog(
	val context: RiskActivity,
	val attacker: Army,
	val defender: Army,
	val attackingDice: IntArray,
	val defendingDice: IntArray,
	val result: BooleanArray
) : Runnable, OnDismissListener, OnCancelListener {
	private val lock = Lock()
	private lateinit var dialog: Dialog
	private lateinit var text: Array<TextView>

	override fun run() {
		val binding = DiceDialogBinding.inflate(LayoutInflater.from(context))
		binding.tvAttacker.text = attacker.prettify()
		binding.tvDefender.text = defender.prettify()
		binding.root.setOnClickListener {
			lock.release()
		}
		val red = arrayOf(
			binding.red1,
			binding.red2,
			binding.red3,
		)
		val white = arrayOf(
			binding.white1,
			binding.white2
		)
		for (i in red.indices) {
			if (i >= attackingDice.size) {
				red[i].visibility = View.INVISIBLE
			} else {
				red[i].rollDice(attackingDice[i], lock)
			}
		}
		for (i in white.indices) {
			if (i >= defendingDice.size)
				white[i].visibility = View.INVISIBLE
			else
				white[i].rollDice(defendingDice[i], lock)
		}
		text = arrayOf(
			binding.text1,
			binding.text2
		)
		dialog = context.newDialogBuilder().setView(binding.root)
			.setNegativeButton("Pause") { _, _ ->
				dialog.dismiss()
				context.game.stopGameThread()
			}.show()
		dialog.setOnCancelListener(this)
		dialog.setCanceledOnTouchOutside(false)
		dialog.setOnDismissListener(this)
	}

	override fun onDismiss(dialog: DialogInterface?) {
		lock.releaseAll()
		context.game.setGameResult(null)
	}

	private fun showResult() {
		for (i in result.indices) {
			text[i].setText(if (result[i]) R.string.arrow_left else R.string.arrow_right)
		}
	}

	override fun onCancel(dialog: DialogInterface?) {
		lock.releaseAll()
		context.game.setGameResult(null)
	}

	init {
		context.runOnUiThread(this)
		Thread.sleep(1000)
		lock.block()
		context.runOnUiThread { showResult() }
		if (dialog.isShowing) {
			lock.acquireAndBlock(6000)
			context.runOnUiThread { dialog.dismiss() }
		}
	}
}